pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	y := make([]byte, length)
+	res := make([]byte, length)
+
+	_, err := crand.Read(x)
+	if err != nil {
+		return nil, err
+	} else if !validateDataIntegrity(x, length) {
+		return nil, errSecureRandomData
+	}
+	_, err = mrand.Read(y)
+	if err != nil {
+		return nil, err
+	} else if !validateDataIntegrity(y, length) {
+		return nil, errSecureRandomData
+	}
+	for i := 0; i < length; i++ {
+		res[i] = x[i] ^ y[i]
+	}
+	if !validateDataIntegrity(res, length) {
+		return nil, errSecureRandomData
+	}
+	return res, nil
+}
+
+func (crypto *defaultCryptoBackend) importECDSAPublic(key *ecdsa.PublicKey) *ecies.PublicKey {
+	return ecies.ImportECDSAPublic(key)
+}
+
+func (crypto *defaultCryptoBackend) GenerateKey() (*ecdsa.PrivateKey, error) {
+	return ethCrypto.GenerateKey()
+}
+
+// CryptoUtils
+
+// NewKeyPair generates a new cryptographic identity for the client, and injects
+// it into the known identities for message decryption. Returns ID of the new key pair.
+func (crypto *defaultCryptoBackend) NewKeyPair(ctx context.Context) (string, error) {
+	key, err := ethCrypto.GenerateKey()
+	if err != nil || !validatePrivateKey(key) {
+		key, err = crypto.GenerateKey() // retry once
+	}
+	if err != nil {
+		return "", err
+	}
+	if !validatePrivateKey(key) {
+		return "", fmt.Errorf("failed to generate valid key")
+	}
+
+	id, err := crypto.generateRandomID()
+	if err != nil {
+		return "", fmt.Errorf("failed to generate ID: %s", err)
+	}
+
+	crypto.keyMu.Lock()
+	defer crypto.keyMu.Unlock()
+
+	if crypto.privateKeys[id] != nil {
+		return "", fmt.Errorf("failed to generate unique ID")
+	}
+	crypto.privateKeys[id] = key
+	return id, nil
+}
+
+func (crypto *defaultCryptoBackend) GetPrivateKey(id string) (*ecdsa.PrivateKey, error) {
+	crypto.keyMu.RLock()
+	defer crypto.keyMu.RUnlock()
+	key := crypto.privateKeys[id]
+	if key == nil {
+		return nil, fmt.Errorf("invalid id")
+	}
+	return key, nil
+}
+
+// Util functions
+
+// validateDataIntegrity returns false if the data have the wrong size or contains all zeros,
+// which is the simplest and the most common bug.
+func validateDataIntegrity(k []byte, expectedSize int) bool {
+	if len(k) != expectedSize {
+		return false
+	}
+	if expectedSize > 3 && containsOnlyZeros(k) {
+		return false
+	}
+	return true
+}
+
+// validatePrivateKey checks the format of the given private key.
+func validatePrivateKey(k *ecdsa.PrivateKey) bool {
+	if k == nil || k.D == nil || k.D.Sign() == 0 {
+		return false
+	}
+	return validatePublicKey(&k.PublicKey)
+}
+
+// ValidatePublicKey checks the format of the given public key.
+func validatePublicKey(k *ecdsa.PublicKey) bool {
+	return k != nil && k.X != nil && k.Y != nil && k.X.Sign() != 0 && k.Y.Sign() != 0
+}
+
+// containsOnlyZeros checks if the data contain only zeros.
+func containsOnlyZeros(data []byte) bool {
+	for _, b := range data {
+		if b != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// bytesToUintLittleEndian converts the slice to 64-bit unsigned integer.
```

Comment:
`binary.LittleEndian.Uint64(bytes)`

Reply:
`binary.LittleEndian.Uint64(bytes)` works different. It expectes 4 bytes. Our encoded sizeField could be 1, 2, 3 or  4 bytes (being 0 the rest of missing bytes).